### PR TITLE
Feature/solution tree

### PR DESCRIPTION
**Pull request by kilativ-dotcom**
solve issue #39 with parameters in solution tree

### CODE REVIEW

**Comment by NikitaZotov** [comment]
Does this feature work?

**Comment by MksmOrlov** [comment]
Also update documentation

**Comment by MksmOrlov** [comment]
Also update documentation and fix typo "applyRile" in file

**Review by NikitaZotov — APPROVED** [review]

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Do not delete pdf. Recompile it with the changes in this PR and push it

**Comment by kilativ-dotcom** [comment]
> Do not delete pdf. Recompile it with the changes in this PR and push it

Deleted pdf by mistake. Now it is back

File: problem-solver/cxx/inferenceModule/generator/SolutionTreeGenerator.cpp [hunk 1]
```diff
@@ -59,19 +60,38 @@ bool SolutionTreeGenerator::addNode(ScAddr const & formula, ScTemplateParams con
...
-ScAddr SolutionTreeGenerator::createSolutionNode(ScAddr const & formula, ScTemplateParams const & templateParams)
+ScAddr SolutionTreeGenerator::createSolutionNode(ScAddr const & formula, ScTemplateParams const & templateParams, std::set<std::string> const & varNames)
 {
-  ScAddr solutionNode = ms_context->CreateNode(ScType::NodeConst);
-  // TODO: Add params to solution node
+  ScAddr const & solutionNode = ms_context->CreateNode(ScType::NodeConst);
   GenerationUtils::generateRelationBetween(ms_context, solutionNode, formula, CoreKeynodes::rrel_1);
+  ScAddr const & replacementsNode = ms_context->CreateNode(ScType::NodeConst);
+  GenerationUtils::generateRelationBetween(ms_context, solutionNode, replacementsNode, CoreKeynodes::rrel_2);
+  for (const auto & varName: varNames)
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/generator/SolutionTreeGenerator.cpp:69]: Do not use auto for primitives (also strings)
```suggestion
  for (std::string const & varName : varNames)
```

File: problem-solver/cxx/inferenceModule/manager/SolutionTreeManager.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    result &= solutionTreeGenerator->addNode(formula, templateParams, varNames);
+  return result;
+}
+
+ScAddr SolutionTreeManager::createSolution(const ScAddr & outputStructure, bool targetAchieved)
+{
+  return solutionTreeGenerator->createSolution(outputStructure, targetAchieved);
+}
+
+bool SolutionTreeManager::checkIfSolutionNodeExists(const ScAddr & rule, ScTemplateParams const & templateParams, std::set<std::string> const & varNames)
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/manager/SolutionTreeManager.cpp:25]:
```suggestion
bool SolutionTreeManager::checkIfSolutionNodeExists(ScAddr const & rule, ScTemplateParams const & templateParams, std::set<std::string> const & varNames)
```

File: problem-solver/cxx/inferenceModule/searcher/SolutionTreeSearcher.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#include "keynodes/InferenceKeynodes.hpp"
+
+namespace inference
+{
+SolutionTreeSearcher::SolutionTreeSearcher(ScMemoryContext * context)
+    : context(context)
+{
+}
+
+bool SolutionTreeSearcher::checkIfSolutionNodeExists(const ScAddr & rule, ScTemplateParams const & templateParams, std::set<std::string> const & varNames)
```
Comment (MksmOrlov) [inline problem-solver/cxx/inferenceModule/searcher/SolutionTreeSearcher.cpp:11]: const after type everywhere
```suggestion
bool SolutionTreeSearcher::checkIfSolutionNodeExists(ScAddr const & rule, ScTemplateParams const & templateParams, std::set<std::string> const & varNames)
```

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Pull request merged**